lib/Reader/readerir.cpp

```diff
@@ -575,6 +576,92 @@ void GenIR::callMonitorHelper(bool IsEnter) {
+  Type *Int32Ty = Type::getInt32Ty(LLVMContext);
+  Type *Int8PtrTy = getUnmanagedPointerType(Int8Ty);
+  IRBuilder<>::InsertPoint SavedInsertPoint = LLVMBuilder->saveIP();
+
+  // Mark this function as requiring a frame pointer and as using GC.
+  Function->addFnAttr("no-frame-pointer-elim-non-leaf");
+  Function->setGC("statepoint-example");
+
+  // The call frame data structure is modeled as an opaque blob of bytes.
+  Type *CallFrameTy = ArrayType::get(Int8Ty, CallFrameInfo.size);
+  Instruction *CallFrameAddress =
+      createTemporary(CallFrameTy, "InlinedCallFrame");
+
+  Type *ThreadPointerTy = getUnmanagedPointerType(ArrayType::get(Int8Ty, 0));
+  Instruction *ThreadPointerAddress =
+      createTemporary(ThreadPointerTy, "ThreadPointer");
+
+  // Intialize the call frame
+  LLVMBuilder->SetInsertPoint(ThreadPointerAddress->getNextNode());
+
+  // Argument 0: &InlinedCallFrame[CallFrameInfo.offsetOfFrameVptr]
+  Value *FrameVPtrIndices[] = {
+      ConstantInt::get(Int32Ty, 0),
+      ConstantInt::get(Int32Ty, CallFrameInfo.offsetOfFrameVptr)};
+  Value *FrameVPtrAddress =
+      LLVMBuilder->CreateInBoundsGEP(CallFrameAddress, FrameVPtrIndices);
+
+  // Argument 1: the secret parameter, if any
+  Value *SecretParam = nullptr;
+  if (MethodSignature.hasSecretParameter()) {
+    SecretParam = (Value *)secretParam();
+  } else {
+    SecretParam = Constant::getNullValue(Int8PtrTy);
+  }
+
+  // Call CORINFO_HELP_INIT_PINVOKE_FRAME
+  const bool MayThrow = false;
+  Value *ThreadPointerValue =
+      callHelperImpl(CORINFO_HELP_INIT_PINVOKE_FRAME, MayThrow, ThreadPointerTy,
+                     (IRNode *)FrameVPtrAddress,
+                     (IRNode *)SecretParam).getInstruction();
+  LLVMBuilder->CreateStore(ThreadPointerValue, ThreadPointerAddress);
+
+  // Store the stack and frame pointers
+  Type *SPType = Type::getIntNTy(LLVMContext, TargetPointerSizeInBits);
+  Type *ReadRegisterTypes[] = {SPType};
+  llvm::Function *ReadRegister = Intrinsic::getDeclaration(
+      JitContext->CurrentModule, Intrinsic::read_register, ReadRegisterTypes);
+  Value *CallSiteSPIndices[] = {
+      ConstantInt::get(Int32Ty, 0),
+      ConstantInt::get(Int32Ty, CallFrameInfo.offsetOfCallSiteSP)};
+  Value *CallSiteSPAddress =
+      LLVMBuilder->CreateInBoundsGEP(CallFrameAddress, CallSiteSPIndices);
+  CallSiteSPAddress = LLVMBuilder->CreatePointerCast(
+      CallSiteSPAddress, getUnmanagedPointerType(SPType));
+  MDString *SPName = MDString::get(LLVMContext, "rsp");
+  Metadata *SPNameNodeMDs[]{SPName};
+  MDNode *SPNameNode = MDNode::get(LLVMContext, SPNameNodeMDs);
+  Value *SPNameValue = MetadataAsValue::get(LLVMContext, SPNameNode);
+  Value *SPValue = LLVMBuilder->CreateCall(ReadRegister, SPNameValue);
+  LLVMBuilder->CreateStore(SPValue, CallSiteSPAddress);
+
+  llvm::Function *FrameAddress = Intrinsic::getDeclaration(
+      JitContext->CurrentModule, Intrinsic::frameaddress);
+  Value *CalleeSavedFPIndices[] = {
```

Comment:
Do you know what the frame address gets used for by the runtime? I wonder if there is some expectation about frame layout we need to honor.